### PR TITLE
Add Zephyr support to TF Lite micro

### PR DESCRIPTION
**Pull request by kgugala**
This PR add support for Zephyr RTOS in TF Lite micro. The PR includes two demos:

* hello_world
* magic_wand

The demos targets sofcore RISC-V implementation running in FPGA. More info and automatic tests can be found in https://github.com/antmicro/litex-vexriscv-tensorflow-lite-demo

### CODE REVIEW

**Comment by kgugala** [comment]
@googlebot I signed it!

**Comment by gbaned** [comment]
@kgugala Thank you for your contribution. Can you please sign CLA? Thanks!

**Comment by kgugala** [comment]
@gbaned done

File: tensorflow/lite/micro/examples/hello_world/zephyr_riscv/CMakeLists.txt
Comment (petewarden) [inline tensorflow/lite/micro/examples/hello_world/zephyr_riscv/CMakeLists.txt:6]: We've been trying to avoid duplicating the definitions of what files to build as our platform support expands, since this can be quite fragile as new files are added to the library. At the moment they only need to be added in one place in the makefile, but with this change they'd need to be added here as well.

The preferred approach is to generate a project file (in this case CMakeLists.txt) as a separate step, driven by the makefile build. You can see some examples of how we do that for Arduino, Mbed, etc here: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/micro/tools/make/helper_functions.inc#L115

Could you try rewriting this PR to use this method?

Comment (kgugala) [reply tensorflow/lite/micro/examples/hello_world/zephyr_riscv/CMakeLists.txt:6]: @petewarden thanks for the review. We'll add a CMakeFiles.txt template and python script which will get the template, list of the source files and generate the final CMakeFiles.txt.

Is that OK?

Comment (kamilrakoczy) [reply tensorflow/lite/micro/examples/hello_world/zephyr_riscv/CMakeLists.txt]: @petewarden I updated PR to generate CMakeLists.txt using `generate_project` from `helper_functions.inc` and it was not necessary to create extra python script. 

I do not copy third_party zephyr folder to generated `prj` folder  as it is large project. To compile generated project standalone, ZEPHYR_BASE environment variable should be set.

Could you review this PR again?

**Review by petewarden — DISMISSED** [review]

**Comment by gbaned** [comment]
@kgugala Can you please fix build failures ? Thanks!

**Comment by mgielda** [comment]
@gbaned can you explain how to do that? I can see that tests on master also fail. I also can't open the "Details" links at least for some of the failures. So what is the criterion we should be looking at?

**Comment by mgielda** [comment]
Or should we just aim for "only those tests fail that also fail on master"? That would probably be doable.

**Comment by gbaned** [comment]
@petewarden Can you please assist on above comments from @kgugala. Thanks!

**Comment by petewarden** [comment]
I looked at the following logs:
https://source.cloud.google.com/results/invocations/62bc0a7c-d21c-4447-86c2-83250d0f0f26/log

In there I see:

```
In file included from ./tensorflow/lite/kernels/internal/reference/reference_ops.h:55:0,
                 from ./tensorflow/lite/kernels/internal/optimized/optimized_ops.h:51,
                 from tensorflow/lite/kernels/split_v.cc:19:
./tensorflow/lite/kernels/internal/reference/reduce.h:23:10: fatal error: tensorflow/lite/kernels/internal/min.h: No such file or directory
 #include "tensorflow/lite/kernels/internal/min.h"
          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
ERROR: /tmpfs/src/github/tensorflow/tensorflow/lite/kernels/BUILD:425:1: Couldn't build file tensorflow/lite/kernels/_objs/builtin_op_kernels/slice.pic.o: C++ compilation of rule
```

This is usually a sign that headers have been added to the repo, but not included in the appropriate Bazel BUILD files. My guess from inspection is that you'll need to add the min.h etc headers to the reference_base target:
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/kernels/internal/BUILD#L431

You can verify this locally by running a `bazel test tensorflow/lite:all`

**Comment by kamilrakoczy** [comment]
@petewarden @gbaned I updated PR and added `min.h` `max.h` to `cppmath` in `tensorflow/lite/kernels/internal/BUILD`, as I think this is better place then `reference_base`. If it is not correct, I can move it to `reference_base`.

**Comment by kgugala** [comment]
@petewarden I had to fix formatting in the BUILD file. Can you restart the kokoro?

**Comment by kamilrakoczy** [comment]
@petewarden Could you restart build again? There was missing `,` in the BUILD file. Now it should be OK.

**Comment by kamilrakoczy** [comment]
@petewarden Could you please restart build once more? Now I installed `buildifier` to ensure BUILD file is correctly formatted.

**Review by petewarden — DISMISSED** [review]

**Comment by mgielda** [comment]
Hi all, just checking up, what else is needed to merge this?

**Comment by kgugala** [comment]
@petewarden @gbaned is there anything we need to address in this PR before it can be merged?

File: tensorflow/lite/micro/examples/magic_wand/zephyr_riscv/src/accelerometer_handler.cc
Comment (rthadur) [inline tensorflow/lite/micro/examples/magic_wand/zephyr_riscv/src/accelerometer_handler.cc:37]: `TF_LITE_REPORT_ERROR should be used instead of error_reporter, so log strings can be removed to save space if needed.`

**Review by rthadur — CHANGES_REQUESTED** [review]
minor nit , can you please change this for all the occurrence in the file.

File: tensorflow/lite/micro/examples/magic_wand/zephyr_riscv/src/accelerometer_handler.cc
Comment (kgugala) [reply tensorflow/lite/micro/examples/magic_wand/zephyr_riscv/src/accelerometer_handler.cc:37]: @rthadur done

**Comment by mgielda** [comment]
I hope that wraps it up -- please review and merge if that's all once the checks complete!

**Review by rthadur — DISMISSED** [review]

**Comment by mgielda** [comment]
What else needs to happen here?

**Comment by rthadur** [comment]
@petewarden can you please help merge this PR ?

**Review by petewarden — DISMISSED** [review]

**Comment by petewarden** [comment]
I took a look at the internal build tests and saw these errors:

```
0508 09:12:19.938 TASK: [418/418] Transform Verify match '(Copyright 20[1-9][0-9]((.|
)*)Apache License, Version 2\.0|Licensed to the Apache Software Foundation)'
0508 09:12:28.998 ERROR: File 'tensorflow/lite/micro/examples/hello_world/zephyr_riscv/prj.conf' failed validation 'Verify match '(Copyright 20[1-9][0-9]((.|
)*)Apache License, Version 2\.0|Licensed to the Apache Software Foundation)''.
0508 09:12:28.998 ERROR: File 'tensorflow/lite/micro/examples/hello_world/zephyr_riscv/src/assert.cc' failed validation 'Verify match '(Copyright 20[1-9][0-9]((.|
)*)Apache License, Version 2\.0|Licensed to the Apache Software Foundation)''.
0508 09:12:28.998 ERROR: File 'tensorflow/lite/micro/examples/magic_wand/zephyr_riscv/boards/litex_vexriscv.overlay' failed validation 'Verify match '(Copyright 20[1-9][0-9]((.|
)*)Apache License, Version 2\.0|Licensed to the Apache Software Foundation)''.
0508 09:12:28.998 ERROR: File 'tensorflow/lite/micro/examples/magic_wand/zephyr_riscv/prj.conf' failed validation 'Verify match '(Copyright 20[1-9][0-9]((.|
)*)Apache License, Version 2\.0|Licensed to the Apache Software Foundation)''.
0508 09:12:28.998 ERROR: File 'tensorflow/lite/micro/examples/magic_wand/zephyr_riscv/src/assert.cc' failed validation 'Verify match '(Copyright 20[1-9][0-9]((.|
)*)Apache License, Version 2\.0|Licensed to the Apache Software Foundation)''.
0508 09:12:29.002 ERROR: 5 file(s) failed the validation of Verify match '(Copyright 20[1-9][0-9]((.|
)*)Apache License, Version 2\.0|Licensed to the Apache Software Foundation)'.
```
Can you add license comments to the start of these files?

**Comment by kamilrakoczy** [comment]
@petewarden I updated PR and added missing copyright headers.

**Review by petewarden — DISMISSED** [review]

**Comment by rthadur** [comment]
@kgugala can you please check this error , it has been failing internally due to this error 

`ERROR: third_party/tensorflow/core/kernels/BUILD:7075:11: Couldn't build file third_party/tensorflow/core/kernels/_objs/portable_tensorflow_kernels/conv_ops_fused_float.o: C++ compilation of rule '//third_party/tensorflow/core/kernels:portable_tensorflow_kernels' failed (Timeout) wrapped_clang failed: error executing command 
(16:41:13) FAILED: Build did NOT complete successfully`

**Comment by kamilrakoczy** [comment]
Hi @rthadur 

I couldn’t find the `//third_party/tensorflow/core/kernels:portable_tensorflow_kernels` target, but I found `//tensorflow/core/kernels:portable_tensorflow_kernels`. I’m assuming `third_party` is just some internal CI folder. Is that correct?

Assuming, the `//tensorflow/core/kernels:portable_tensorflow_kernels` target is correct, could you provide more details about the fail. To be more precise - which configuration of the target is failing?

I was able to build the `//tensorflow/core/kernels:portable_tensorflow_kernels` target on my local machine. What I noticed was that when building the target for Android configuration it consumes a lot of memory. Building with `-j12` used over 30GB of RAM and caused the build process to be killed by the OOM killer. However, `-j6` worked perfectly fine. Could this be a reason for the failures in CI?

**Comment by kgugala** [comment]
@petewarden @rthadur We have rebased this PR on top of current master.

Also, we noticed that the build system complains about overriding recipes. We fixed that in #39901

**Review by petewarden — DISMISSED** [review]

**Comment by rthadur** [comment]
@kgugala here is the internal we are seeing , can you please check this 

`file included from tensorflow/lite/micro/kernels/reduce.cpp:16:0:
tensorflow/lite/kernels/internal/reference/reduce.h:21:10: fatal error: tensorflow/lite/kernels/internal/max.h: No such file or directory
 #include "tensorflow/lite/kernels/internal/max.h"
          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
Multiple libraries were found for "TensorFlowLite.h"
Error: build failed: exit status 1`

**Comment by kgugala** [comment]
@rthadur the failing file `tensorflow/lite/micro/kernels/reduce.cpp` does not seem to exist in the public repo https://github.com/tensorflow/tensorflow/tree/master/tensorflow/lite/micro/kernels (it doesn't seem to be present in the repo history). Can you point us to the exact version of the TF you're testing?

**Comment by petewarden** [comment]
This is actually an issue with the Arduino build process, here's a more full version of the error:

```
In file included from /root/Arduino/libraries/tensorflow_lite/src/tensorflow/lite/micro/kernels/reduce.cpp:16:0:
/root/Arduino/libraries/tensorflow_lite/src/tensorflow/lite/kernels/internal/reference/reduce.h:21:10: fatal error: tensorflow/lite/kernels/internal/max.h: No such file or directory
 #include "tensorflow/lite/kernels/internal/max.h"
          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
```

You can run this automated test build yourself with the following script:

https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/micro/tools/ci_build/test_arduino.sh

From looking at the changes, I think you need to add the min/max.h files to this list in the makefile:
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/micro/tools/make/Makefile#L187

Thanks for your patience with this process, hopefully we're nearly there!

**Comment by kgugala** [comment]
@petewarden thanks! Should be fixed now.

**Review by rthadur — APPROVED** [review]

**Review by rthadur — APPROVED** [review]

**Comment by rthadur** [comment]
Changes got submitted internally, waiting for auto-merge to happen. Thank you

**Pull request merged**
